glean-core/src/event_database/mod.rs

```diff
@@ -238,34 +311,216 @@ impl EventDatabase {
+                        execution_counter, cur_ec
+                    ),
+                    None,
+                );
+                // Let's fix cur_ec up and hope this isn't a sign something big is broken.
+                cur_ec = execution_counter;
+            }
+            if highest_ts > event.timestamp {
+                // Even though we sorted everything, something in the
+                // execution_counter or glean.startup.date math went awry.
+                record_error(
+                    glean,
+                    &glean_restarted_meta(store_name),
+                    ErrorType::InvalidState,
+                    format!(
+                        "Inconsistent previous highest timestamp {} (expected <= {})",
+                        highest_ts, event.timestamp
+                    ),
+                    None,
+                );
+                // Let the highest_ts regress to event.timestamp to hope this minimizes weirdness.
+            }
+            highest_ts = event.timestamp
+        }
+    }
+
     /// Gets a snapshot of the stored event data as a JsonValue.
     ///
     /// # Arguments
     ///
+    /// * `glean` - the Glean instance.
     /// * `store_name` - The name of the desired store.
     /// * `clear_store` - Whether to clear the store after snapshotting.
     ///
     /// # Returns
     ///
     /// A array of events, JSON encoded, if any. Otherwise `None`.
-    pub fn snapshot_as_json(&self, store_name: &str, clear_store: bool) -> Option<JsonValue> {
+    pub fn snapshot_as_json(
+        &self,
+        glean: &Glean,
+        store_name: &str,
+        clear_store: bool,
+    ) -> Option<JsonValue> {
         let result = {
             let mut db = self.event_stores.write().unwrap(); // safe unwrap, only error case is poisoning
             db.get_mut(&store_name.to_string()).and_then(|store| {
                 if !store.is_empty() {
-                    // Timestamps may have been recorded out-of-order, so sort the events
-                    // by the timestamp.
-                    // We can't insert events in order as-we-go, because we also append
-                    // events to a file on disk, where this would be expensive. Best to
-                    // handle this in every case (whether events came from disk or memory)
-                    // in a single location.
-                    store.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));
-                    let first_timestamp = store[0].timestamp;
-                    let snapshot = store
-                        .iter()
-                        .map(|e| e.serialize_relative(first_timestamp))
-                        .collect();
-                    Some(snapshot)
+                    // Normalization happens in-place, so if we're not clearing,
+                    // operate on a clone.
+                    if !clear_store {
+                        *store = store.clone();
```

Comment:
Doesn't this overwrite the data we have in memory and operates on that?
You're cloning the data, but then write it in the same place it came from.

Reply:
You're right. I think I need a `let store =` in here.